load-generator-source/src/main/java/org/springframework/cloud/stream/module/loadgenerator/LoadGeneratorSource.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.module.loadgenerator;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.stream.annotation.EnableBinding;
+import org.springframework.cloud.stream.messaging.Source;
+import org.springframework.integration.endpoint.AbstractEndpoint;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.support.GenericMessage;
+
+/**
+ * A source that sends a set amount of empty byte array messages to verify the speed
+ * of the infrastructure.
+ *
+ * @author Glenn Renfro
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties({LoadGeneratorSourceProperties.class})
+public class LoadGeneratorSource extends AbstractEndpoint {
+
+	private static final Logger logger = LoggerFactory.getLogger(LoadGeneratorSource.class);
+
+	@Autowired
+	private volatile LoadGeneratorSourceProperties config;
```

Comment:
This is a good question! It depends on whether it is guaranteed that the same thread that performs the auto wiring is also the same thread that invokes `doStart()`. If so, then we don't need these fields to be `volatile`.

However, if a separate thread could invoke `doStart()` then we have to determine if updates to those fields will be visible to that thread.

So far I'm seeing conflicting information. According to this, it seems that we can do without `volatile` since the fields are only set once:

> Changes to fields made by one thread are guaranteed to be visible to other threads only under the following conditions:
> ...
> The first time a thread accesses a field of an object, it sees either the initial value of the field or a value since written by some other thread.
>
> http://gee.cs.oswego.edu/dl/cpj/jmm.html

On the other hand, this initialization pattern is literally the textbook definition of unsafe publication according to _Java Concurrency in Practice_:

![screen shot 2015-11-12 at 11 10 32 am](https://cloud.githubusercontent.com/assets/394956/11123685/10a65b30-892f-11e5-99f9-53cb398dc474.png)
I'll investigate some more and post an update shortly.